File: src/geometry/wm/convert/flat.ts
import { extendBBox } from '../../bbox.js';
import { s2ToVector } from '../../index.js';

import type {
  BBOX,
  Coordinates,
  Feature,
  Geometry,
  GeometryType,
  LineString,
  LineString3D,
  LineStringMValues,
  MValue,
  MValues,
  MultiLineString,
  MultiLineString3D,
  MultiLineStringMValues,
  MultiPolygon,
  MultiPolygon3D,
  MultiPolygonMValues,
  Point,
  Point3D,
  Polygon3D,
  Properties,
  VectorFeatures,
  VectorGeometry,
  VectorLineString,
  VectorMultiLineString,
  VectorMultiPolygon,
  VectorPoint,
} from '../../index.js';

/**
 * Convert a GeoJSON Vector Feature to a GeoJSON Feature
 * @param data - GeoJSON Vector Feature
 * @param buildBBox - optional - build a bbox for the feature if desired
 * @returns - GeoJson Feature
 */
export function toFlatGeometry<
  M = Record<string, unknown>,
  D extends MValue = Properties,
  P extends Properties = Properties,
  G extends VectorGeometry<D> = VectorGeometry<D>,
>(data: VectorFeatures<M, D, P, G>, buildBBox?: boolean): Feature<M, D, P, Geometry<D>> {
  const { id, type, face, properties, metadata } = data;
  const vectorGeo = data.geometry;
  // adjust S2 to Vector if needed
  if (type === 'S2Feature') s2ToVector<D>(face ?? 0, data.geometry);
  const flatGeo = vectorToFlat<D>(vectorGeo, buildBBox);
  return {
    id,
    type: 'Feature',
    properties,
    metadata,
    geometry: flatGeo,
  };
}

/**
 * Convert a GeoJSON Vector Geometry to a Flat GeoJSON Geometry
 * @param geometry - GeoJSON Vector Geometry
 * @param buildBBox - optional - build a bbox for the feature if desired
 * @returns - GeoJSON Geometry
 */
export function vectorToFlat<M extends MValue = Properties>(
  geometry: VectorGeometry<M>,
  buildBBox?: boolean,
): Geometry<M> {
  const { type: vType, coordinates: coords, bbox, is3D } = geometry;
  const newBBox: BBOX | undefined =
    buildBBox === true && bbox === undefined ? ([] as unknown as BBOX) : undefined;

  let res: [Coordinates, MValue | MValues | undefined];
  if (vType === 'Point') res = convertPoint(coords, newBBox);
  else if (vType === 'MultiPoint' || vType === 'LineString')
    res = convertLineString(coords, newBBox);
  else if (vType === 'MultiLineString' || vType === 'Polygon')
    res = convertMultiLineString(coords, newBBox);
  else if (vType === 'MultiPolygon') res = convertMultiPolygon(coords, newBBox);
  else {
    throw new Error('Invalid Geometry type');
  }

  const [coordinates, mValues] = res;
  const type: GeometryType = is3D ? `${vType}3D` : vType;
  // @ts-expect-error - type complains but it works
  return { type, coordinates, bbox: newBBox ?? bbox, mValues };
}

/**
 * Convert a GeoJSON Vector MultiPolygon to a GeoJSON MultiPolygon
 * @param input - GeoJSON Vector MultiPolygon
 * @param bbox - if bbox is provided, we will extend the bbox
 * @returns - GeoJSON MultiPolygon
 */
function convertMultiPolygon(
  input: VectorMultiPolygon,
  bbox?: BBOX,
): [MultiPolygon | MultiPolygon3D, MultiPolygonMValues | undefined] {
  const polygons: MultiPolygon3D = [];
  const mValues: MultiPolygonMValues = [];
  let mValuesFound = false;
  for (const polygon of input) {
    const [poly, mValue] = convertMultiLineString(polygon, bbox);
    polygons.push(poly as unknown as Polygon3D);
    if (mValue !== undefined) mValuesFound = true;
    mValues.push(mValue ?? []);
  }
  return [polygons, mValuesFound ? mValues : undefined];
}

/**
 * Convert a GeoJSON Vector MultiLineString to a GeoJSON MultiLineString
 * @param input - GeoJSON Vector MultiLineString
 * @param bbox - if bbox is provided, we will extend the bbox
 * @returns - GeoJSON MultiLineString
 */
function convertMultiLineString(
  input: VectorMultiLineString,
  bbox?: BBOX,
): [MultiLineString | MultiLineString3D, MultiLineStringMValues | undefined] {
  const lines: MultiLineString3D = [];
  const mValues: MultiLineStringMValues = [];
  let mValuesFound = false;
  for (const line of input) {
    const [lineString, mValue] = convertLineString(line, bbox);
    lines.push(lineString as unknown as LineString3D);
    if (mValue !== undefined) mValuesFound = true;
    mValues.push(mValue ?? []);
  }
  return [lines, mValuesFound ? mValues : undefined];
}

/**
 * Convert a GeoJSON Vector LineString to a GeoJSON LineString
 * @param input - GeoJSON Vector LineString
 * @param bbox - if bbox is provided, we will extend the bbox
 * @returns - GeoJSON LineString
 */
function convertLineString(
  input: VectorLineString,
  bbox?: BBOX,
): [LineString | LineString3D, LineStringMValues | undefined] {
  const lines: LineString3D = [];
  const mValues: LineStringMValues = [];
  let mValuesFound = false;
  for (const vp of input) {
    const [point, mValue] = convertPoint(vp, bbox);
    lines.push(point as unknown as Point3D);
    if (mValue !== undefined) mValuesFound = true;
    mValues.push(mValue ?? {});
  }
  return [lines, mValuesFound ? mValues : undefined];
}

/**
 * Mutate a GeoJSON Vector Point to a GeoJSON Point or Point3D
 * @param point - GeoJSON Vector Point
 * @param bbox - if bbox is provided, we will extend the bbox
 * @returns - GeoJSON Point or Point3D
 */
function convertPoint(point: VectorPoint, bbox?: BBOX): [Point | Point3D, MValue | undefined] {
  if (bbox !== undefined) {
    const newBBox = extendBBox(bbox, point);
    for (let i = 0; i < newBBox.length; i++) bbox[i] = newBBox[i];
  }
  if (point.z === undefined) return [[point.x, point.y], point.m];
  return [[point.x, point.y, point.z], point.m];
}

File: rust/geometry/wm/convert/flat.rs
use s2json::{
    BBox3D, Geometry, GeometryType, LineString3DGeometry, LineStringGeometry,
    MultiLineString3DGeometry, MultiLineStringGeometry, MultiPoint3DGeometry, MultiPointGeometry,
    MultiPolygon3DGeometry, MultiPolygonGeometry, Point, Point3D, Point3DGeometry, PointGeometry,
    Polygon3DGeometry, PolygonGeometry, VectorGeometry, VectorLineStringGeometry,
    VectorMultiLineStringGeometry, VectorMultiPointGeometry, VectorMultiPolygonGeometry,
    VectorPointGeometry, VectorPolygonGeometry,
};

/// Convert a GeoJSON Geometry to an Vector Geometry
pub fn convert_vector_to_geometry<M: Clone + Default>(
    geometry: &VectorGeometry<M>,
    build_bbox: bool,
) -> Geometry<M> {
    let bbox = if build_bbox { Some(BBox3D::default()) } else { None };
    match geometry {
        VectorGeometry::Point(geo) => to_point(geo, bbox),
        VectorGeometry::MultiPoint(geo) => to_points(geo, bbox),
        VectorGeometry::LineString(geo) => to_linestring(geo, bbox),
        VectorGeometry::MultiLineString(geo) => to_linestrings(geo, bbox),
        VectorGeometry::Polygon(geo) => to_polygon(geo, bbox),
        VectorGeometry::MultiPolygon(geo) => to_polygons(geo, bbox),
    }
}

/// Mutate a GeoJSON Vector Point to a GeoJSON Point or Point3D
fn to_point<M: Clone + Default>(
    point: &VectorPointGeometry<M>,
    mut bbox: Option<BBox3D>,
) -> Geometry<M> {
    let coords = &point.coordinates;
    if let Some(b) = bbox.as_mut() {
        b.extend_from_point(coords);
    }
    if point.is_3d {
        Geometry::Point3D(Point3DGeometry::<M> {
            _type: GeometryType::Point3D,
            coordinates: Point3D(coords.x, coords.y, coords.z.unwrap_or_default()),
            m_values: coords.m.clone(),
            bbox,
        })
    } else {
        Geometry::Point(PointGeometry::<M> {
            _type: GeometryType::Point,
            coordinates: Point(coords.x, coords.y),
            m_values: coords.m.clone(),
            bbox: bbox.map(|b| b.into()),
        })
    }
}

/// Mutate a GeoJSON Vector MultiPoint to a GeoJSON MultiPoint/MultiPoint3D
fn to_points<M: Clone + Default>(
    points: &VectorMultiPointGeometry<M>,
    mut bbox: Option<BBox3D>,
) -> Geometry<M> {
    let coords = &points.coordinates;
    if let Some(b) = bbox.as_mut() {
        coords.iter().for_each(|p| b.extend_from_point(p));
    }
    let m_values = coords.iter().map(|p| p.m.clone()).collect();
    if points.is_3d {
        Geometry::MultiPoint3D(MultiPoint3DGeometry::<M> {
            _type: GeometryType::MultiPoint3D,
            coordinates: coords
                .iter()
                .map(|p| Point3D(p.x, p.y, p.z.unwrap_or_default()))
                .collect(),
            m_values,
            bbox,
        })
    } else {
        Geometry::MultiPoint(MultiPointGeometry::<M> {
            _type: GeometryType::MultiPoint,
            coordinates: coords.iter().map(|p| Point(p.x, p.y)).collect(),
            m_values,
            bbox: bbox.map(|b| b.into()),
        })
    }
}

/// Mutate a GeoJSON Vector LineString to a GeoJSON LineString/LineString3D
fn to_linestring<M: Clone + Default>(
    linestring: &VectorLineStringGeometry<M>,
    mut bbox: Option<BBox3D>,
) -> Geometry<M> {
    let coords = &linestring.coordinates;
    if let Some(b) = bbox.as_mut() {
        coords.iter().for_each(|p| b.extend_from_point(p));
    }
    let m_values = coords.iter().map(|p| p.m.clone()).collect();
    if linestring.is_3d {
        Geometry::LineString3D(LineString3DGeometry {
            _type: GeometryType::LineString3D,
            coordinates: coords
                .iter()
                .map(|p| Point3D(p.x, p.y, p.z.unwrap_or_default()))
                .collect(),
            m_values,
            bbox,
        })
    } else {
        Geometry::LineString(LineStringGeometry {
            _type: GeometryType::LineString,
            coordinates: coords.iter().map(|p| Point(p.x, p.y)).collect(),
            m_values,
            bbox: bbox.map(|b| b.into()),
        })
    }
}

/// Mutate a GeoJSON Vector MultiLineString to a GeoJSON MultiLineString/MultiLineString3D
fn to_linestrings<M: Clone + Default>(
    linestrings: &VectorMultiLineStringGeometry<M>,
    mut bbox: Option<BBox3D>,
) -> Geometry<M> {
    let coords = &linestrings.coordinates;
    if let Some(b) = bbox.as_mut() {
        coords.iter().for_each(|l| l.iter().for_each(|p| b.extend_from_point(p)));
    }
    let m_values = coords.iter().map(|l| l.iter().map(|p| p.m.clone()).collect()).collect();
    if linestrings.is_3d {
        Geometry::MultiLineString3D(MultiLineString3DGeometry {
            _type: GeometryType::MultiLineString3D,
            coordinates: coords
                .iter()
                .map(|l| l.iter().map(|p| Point3D(p.x, p.y, p.z.unwrap_or_default())).collect())
                .collect(),
            m_values,
            bbox,
        })
    } else {
        Geometry::MultiLineString(MultiLineStringGeometry {
            _type: GeometryType::MultiLineString,
            coordinates: coords
                .iter()
                .map(|l| l.iter().map(|p| Point(p.x, p.y)).collect())
                .collect(),
            m_values,
            bbox: bbox.map(|b| b.into()),
        })
    }
}

/// Mutate a GeoJSON Vector Polygon to a GeoJSON Polygon/Polygon3D
fn to_polygon<M: Clone + Default>(
    polygon: &VectorPolygonGeometry<M>,
    mut bbox: Option<BBox3D>,
) -> Geometry<M> {
    let coords = &polygon.coordinates;
    if let Some(b) = bbox.as_mut() {
        coords.iter().for_each(|r| r.iter().for_each(|p| b.extend_from_point(p)));
    }
    let m_values = coords.iter().map(|r| r.iter().map(|p| p.m.clone()).collect()).collect();
    if polygon.is_3d {
        Geometry::Polygon3D(Polygon3DGeometry {
            _type: GeometryType::Polygon3D,
            coordinates: coords
                .iter()
                .map(|r| r.iter().map(|p| Point3D(p.x, p.y, p.z.unwrap_or_default())).collect())
                .collect(),
            m_values,
            bbox,
        })
    } else {
        Geometry::Polygon(PolygonGeometry {
            _type: GeometryType::Polygon,
            coordinates: coords
                .iter()
                .map(|r| r.iter().map(|p| Point(p.x, p.y)).collect())
                .collect(),
            m_values,
            bbox: bbox.map(|b| b.into()),
        })
    }
}

/// Mutate a GeoJSON Vector MultiPolygon to a GeoJSON MultiPolygon/MultiPolygon3D
fn to_polygons<M: Clone + Default>(
    polygons: &VectorMultiPolygonGeometry<M>,
    mut bbox: Option<BBox3D>,
) -> Geometry<M> {
    let coords = &polygons.coordinates;
    if let Some(b) = bbox.as_mut() {
        coords
            .iter()
            .for_each(|p| p.iter().for_each(|r| r.iter().for_each(|p| b.extend_from_point(p))));
    }
    let m_values = coords
        .iter()
        .map(|p| p.iter().map(|r| r.iter().map(|p| p.m.clone()).collect()).collect())
        .collect();
    if polygons.is_3d {
        Geometry::MultiPolygon3D(MultiPolygon3DGeometry {
            _type: GeometryType::MultiPolygon3D,
            coordinates: coords
                .iter()
                .map(|p| {
                    p.iter()
                        .map(|r| {
                            r.iter().map(|p| Point3D(p.x, p.y, p.z.unwrap_or_default())).collect()
                        })
                        .collect()
                })
                .collect(),
            m_values,
            bbox,
        })
    } else {
        Geometry::MultiPolygon(MultiPolygonGeometry {
            _type: GeometryType::MultiPolygon,
            coordinates: coords
                .iter()
                .map(|p| p.iter().map(|r| r.iter().map(|p| Point(p.x, p.y)).collect()).collect())
                .collect(),
            m_values,
            bbox: bbox.map(|b| b.into()),
        })
    }
}
